### ISSUE TITLE
Can we add necessary labels when volume controller annotate PVC?

### ISSUE THREAD

**Issue by cvvz** [kubernetes/kubernetes#127426]
### What would you like to be added?

Currently, volume controller will add `volume.kubernetes.io/storage-provisioner` annotation to the PVC to inform specific external-provisioner to provision the volume.  

However, external-provisioner list and watch PVC objects without any filter, that means it will save all the PVC objects in memory, not only the ones should be handled by it. This will cause oom when there are too many PVCs in the cluster, even though the PVCs are not to be handled by the external-provisioner. 

Can we add labels when volume controller add `volume.kubernetes.io/storage-provisioner` annotation to the PVC and add label filter in external-provisioner's informer to only list-watch the PVCs that should be provisioned by it?

### Why is this needed?

external-provisioner will OOM when there are too many PVCs in the cluster, even the PVCs should not be handled by it. We can use label-filter to avoid saving all the PVCs in memory.

**Comment by andyzhangx** [comment]
cc @jsafrane @msau42

**Comment by thisisharrsh** [comment]
I agree that adding labels and utilizing label filtering in external-provisioner's informer could greatly improve memory usage, especially in large clusters.

**Comment by thisisharrsh** [comment]
I have a thought:
Should we make the label addition configurable for flexibility, or should it always be applied when the `volume.kubernetes.io/storage-provisioner` annotation is added?

**Comment by jsafrane** [comment]
I like this. 

> Should we make the label addition configurable for flexibility, or should it always be applied when the volume.kubernetes.io/storage-provisioner annotation is added?

I think pv-controller can *always* add it when it adds the annotation. And it should add the label when it sees the annotation already exists (added by old KCM0. We would need to have it optional in the external-provisioner for a while, to be able to handle clusters with old KCMs that do not label PVs.

**Comment by jsafrane** [comment]
Oh, and we would need to be careful with CSI migration. Just test that PVCs get the right label that the external-provisioner understands, even if the StorageClass refers to in-tree volume plugin.

**Comment by thisisharrsh** [comment]
> We would need to have it optional in the external-provisioner for a while, to be able to handle clusters with old KCMs that do not label PVs.



> we would need to be careful with CSI migration.

I totally agree with your points!

**Comment by thisisharrsh** [comment]
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/pv_controller.go#L237 

We can have a logic for an annotation check in the `syncClaim(ctx context.Context, claim *v1.PersistentVolumeClaim)` function, and if the annotation exists, we can add a label.

**Comment by thisisharrsh** [comment]
Is this approach good? 
It would be great if some other contributors could share their thoughts.
